content/blog/peace-of-mind-with-cloud-secret-providers/index.md

```diff
@@ -0,0 +1,669 @@
+```csharp
+using Pulumi;
+using Kms = Pulumi.Aws.Kms;
+
+class KeyStack : Stack
+{
+    public KeyStack()
+    {
+        var config = new Config();
+        var iamRole = config.Require("iamRole");
+        var accountID = config.Require("accountID");
+
+        var keyPolicy = $@"{{
+            ""Version"": ""2012-10-17"",
+            ""Id"": ""policy"",
+            ""Statement"": [
+            {{
+                ""Sid"": ""AllowGetKeys"",
+                ""Effect"": ""Allow"",
+                ""Action"": [""kms:Describe*"", ""kms:Get*"", ""kms:List*""],
+                ""Principal"": {{
+                    ""AWS"": [""arn:aws:iam::{accountID}:root""]
+                }},
+                ""Resource"": ""*""
+            }},
+            {{
+                ""Sid"": ""AllowIAMUserAccessKeys"",
+                ""Effect"": ""Allow"",
+                ""Action"": [""kms:*""],
+                ""Principal"": {{
+                    ""AWS"": ""{iamRole}""
+                }},
+                ""Resource"": ""*""
+            }}
+            ]
+        }}";
+
+        // Create a new KMS key
+        var key = new Kms.Key("stack-encryption-key", new Kms.KeyArgs
+        {
+            DeletionWindowInDays = 10,
+            Description = "KMS key for encrypting Pulumi secret values",
+            Policy = keyPolicy
+        });
+
+        // Create a new alias to the key
+        var alias = new Kms.Alias("alias/stack-encryption-key", new Kms.AliasArgs
+        {
+            TargetKeyId = key.KeyId
+        });
+
+        this.KeyArn = key.Arn;
+        this.AliasArn = alias.Arn;
+    }
+
+    [Output("keyArn")] public Output<string> KeyArn { get; set; }
+    [Output("aliasArn")] public Output<string> AliasArn { get; set; }
+}
+```
+{{% /choosable %}}
+
+{{< /chooser >}}
+
+In this example, we've made the IAM role allowed access to the key in the key policy configurable to an IAM role, and we've also given read permissions to the key to everyone in the account. Let's set some configuration variables so we can be sure we can use this key for our next Pulumi stack:
+
+```bash
+# Get the current AWS account ID and set it as a config variable
+aws sts get-caller-identity | jq .Account -r | pulumi config set <projectname>:accountID
+# Get the current IAM role we're using as set it as a config variable
+aws sts get-caller-identity | jq .Arn -r | pulumi config set <projectname>:iamRole
+```
+
+## Initialize a New Stack
+
+Now our key has been created and is adequately scoped; we can create a new stack and use the `secrets-provider` flag on creation to specify the KMS key to encrypt our secrets.
+
+```bash
+# In our examples we set our alias to "stack-encryption-key" so we'll use that
+# Note, we need to set the region we're deploying to
+pulumi new aws-typescript -n <projectname> -s <stackname> -d "An example stack encrypted with AWS KMS" --secrets-provider="awskms://alias/stack-encryption-key?region=us-west-2" --config aws:region=us-west-2
```

Comment:
Does `stack-encryption-key` not end up having any random hex at the end in practice?  I might have expected a `pulumi stack output` to be needed here?